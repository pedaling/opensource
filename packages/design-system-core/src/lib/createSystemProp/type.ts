import type { Theme } from '../types';

export type InterpolationFn = (props: Record<string, any>) => Record<string, any>;

export type SystemPropConfig = {
  property: string;
  styleProperty?: string;
  scale?: keyof Pick<Theme, 'colors' | 'space' | 'opacity'>;
  disabled?: boolean;
  transform?: (value: any) => Record<string, any> | undefined;
};

export type SystemProp = {
  (props: Record<string, any>): Record<string, any>;
  propName: string;
  disabled: boolean;
};
