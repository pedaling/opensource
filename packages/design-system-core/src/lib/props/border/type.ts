import type { ResponsiveValue } from '../../types';

export type BorderProps = {
  borderWidth?: ResponsiveValue<number>;
  borderStyle?: ResponsiveValue<'solid' | 'none'>;
  borderColor?: ResponsiveValue<string>;
  borderTopWidth?: ResponsiveValue<number>;
  borderTopStyle?: ResponsiveValue<'solid' | 'none'>;
  borderTopColor?: ResponsiveValue<string>;
  borderRightWidth?: ResponsiveValue<number>;
  borderRightStyle?: ResponsiveValue<'solid' | 'none'>;
  borderRightColor?: ResponsiveValue<string>;
  borderBottomWidth?: ResponsiveValue<number>;
  borderBottomStyle?: ResponsiveValue<'solid' | 'none'>;
  borderBottomColor?: ResponsiveValue<string>;
  borderLeftWidth?: ResponsiveValue<number>;
  borderLeftStyle?: ResponsiveValue<'solid' | 'none'>;
  borderLeftColor?: ResponsiveValue<string>;
  borderRadius?: ResponsiveValue<number | string>;
  borderTopLeftRadius?: ResponsiveValue<number | string>;
  borderTopRightRadius?: ResponsiveValue<number | string>;
  borderBottomLeftRadius?: ResponsiveValue<number | string>;
  borderBottomRightRadius?: ResponsiveValue<number | string>;
  outlineWidth?: ResponsiveValue<number>;
  outlineStyle?: ResponsiveValue<'solid' | 'none'>;
  outlineColor?: ResponsiveValue<string>;
  outlineOffset?: ResponsiveValue<number>;
};
