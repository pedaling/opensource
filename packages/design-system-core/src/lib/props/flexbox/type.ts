import type { ResponsiveValue } from '../../types';

export type FlexboxProps = {
  flex?: ResponsiveValue<string | number>;
  flexGrow?: ResponsiveValue<string | number>;
  flexShrink?: ResponsiveValue<string | number>;
  flexBasis?: ResponsiveValue<string | number>;
  flexDirection?: ResponsiveValue<'column' | 'column-reverse' | 'row' | 'row-reverse'>;
  flexWrap?: ResponsiveValue<'nowrap' | 'wrap' | 'wrap-reverse'>;
  alignContent?: ResponsiveValue<
    | 'space-around'
    | 'space-between'
    | 'space-evenly'
    | 'stretch'
    | 'center'
    | 'end'
    | 'flex-end'
    | 'flex-start'
    | 'start'
    | 'baseline'
    | 'normal'
  >;
  alignItems?: ResponsiveValue<
    | 'center'
    | 'end'
    | 'flex-end'
    | 'flex-start'
    | 'self-end'
    | 'self-start'
    | 'start'
    | 'baseline'
    | 'normal'
    | 'stretch'
  >;
  alignSelf?: ResponsiveValue<
    | 'center'
    | 'end'
    | 'flex-end'
    | 'flex-start'
    | 'self-end'
    | 'self-start'
    | 'start'
    | 'auto'
    | 'baseline'
    | 'normal'
    | 'stretch'
  >;
  justifyContent?: ResponsiveValue<
    | 'space-around'
    | 'space-between'
    | 'space-evenly'
    | 'stretch'
    | 'center'
    | 'end'
    | 'flex-end'
    | 'flex-start'
    | 'start'
    | 'left'
    | 'normal'
    | 'right'
  >;
  justifyItems?: ResponsiveValue<
    | 'center'
    | 'end'
    | 'flex-end'
    | 'flex-start'
    | 'self-end'
    | 'self-start'
    | 'start'
    | 'baseline'
    | 'left'
    | 'legacy'
    | 'normal'
    | 'right'
    | 'stretch'
  >;
  justifySelf?: ResponsiveValue<
    | 'center'
    | 'end'
    | 'flex-end'
    | 'flex-start'
    | 'self-end'
    | 'self-start'
    | 'start'
    | 'auto'
    | 'baseline'
    | 'left'
    | 'normal'
    | 'right'
    | 'stretch'
  >;
  order?: ResponsiveValue<number>;
};
