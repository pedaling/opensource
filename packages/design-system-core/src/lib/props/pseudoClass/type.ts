import type { ResponsiveValue } from '@class101/design-system-core';
import type { SystemProps } from '../systemProp';

export type PseudoClassProps = {
  pseudoHover?: ResponsiveValue<SystemProps>;
  pseudoFocus?: ResponsiveValue<SystemProps>;
  pseudoActive?: ResponsiveValue<SystemProps>;
};
