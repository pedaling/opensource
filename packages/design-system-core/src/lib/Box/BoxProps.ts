import type { Ref, ComponentType, ReactNode } from 'react';
import { createShouldForwardProp } from '../createShouldForwardProp';
import type { SystemProp } from '../createSystemProp';
import type { SystemProps } from '../props';
import { systemProps } from '../props';

export type BoxProps<
  BaseComponent extends ComponentType | unknown = unknown,
  ElementName extends keyof JSX.IntrinsicElements = 'div',
  ElementProps extends JSX.IntrinsicElements[ElementName] = JSX.IntrinsicElements[ElementName]
> = (BaseComponent extends ComponentType<infer BaseComponentProps> ? BaseComponentProps : Record<never, never>) &
  SystemProps & {
    as?: ElementName;
    base?: BaseComponent;
    id?: string;
    ref?: Ref<BaseComponent extends abstract new (...args: any) => any ? InstanceType<BaseComponent> : HTMLElement>;
    children?: ReactNode;
  } & Omit<ElementProps, keyof SystemProps>;

const propNames = systemProps.filter(systemProp => !systemProp.disabled).map(systemProp => systemProp.propName);

export const shouldForwardProp = createShouldForwardProp(propNames);

const systemPropCache: Record<string, SystemProp | null> = {};

const isObject = (obj: Record<string, any>) => typeof obj === 'object' && obj !== null;

export const interpolation = (props: Record<string, any>): Record<string, any> => {
  let result = {};

  if (!isObject(props)) {
    return props;
  }

  if (Array.isArray(props)) {
    return props.map(prop => interpolation(prop));
  }

  for (const [key, value] of Object.entries(props)) {
    const matchedSystemProp =
      systemPropCache[key] === undefined
        ? systemProps.find(systemProp => systemProp.propName === key)
        : systemPropCache[key];

    if (!matchedSystemProp) {
      systemPropCache[key] = null;
      continue;
    }

    systemPropCache[key] ||= matchedSystemProp;

    const prop = {
      [key]: interpolation(value),
    };

    result = mergeMediaQueries(result, matchedSystemProp(prop));
  }

  return result;
};

const mergeMediaQueries = (original: Record<string, any>, next: Record<string, any>) => {
  const mediaQueryProperties: Record<string, any> = {};

  for (const key of Object.keys(next)) {
    if (key.startsWith('@media')) {
      mediaQueryProperties[key] = { ...(original[key] ?? {}), ...next[key] };
    }
  }

  return { ...original, ...next, ...mediaQueryProperties };
};
