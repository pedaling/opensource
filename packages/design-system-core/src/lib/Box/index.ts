export type { BoxProps } from './BoxProps';
export { Box } from './Box';
