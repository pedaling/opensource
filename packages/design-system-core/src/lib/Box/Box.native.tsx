import type { ComponentType, FC } from 'react';
import { forwardRef } from 'react';
import { Text, TextInput, TouchableOpacity, View } from 'react-native';
import styled from '@emotion/native';
import type { BoxProps } from './BoxProps';
import { shouldForwardProp, interpolation } from './BoxProps';

const transformAs = (as: keyof JSX.IntrinsicElements) => {
  switch (as) {
    case 'button':
      return TouchableOpacity;
    case 'input':
      return TextInput;
    case 'span':
    case 'a':
    case 'p':
    case 'h1':
    case 'h2':
    case 'h3':
    case 'h4':
    case 'h5':
      return Text;
    default:
      return View;
  }
};

export const Box: FC<BoxProps> = styled(
  forwardRef<HTMLElement, BoxProps>(({ as, base, ...restProps }, ref) => {
    const Component = base ? (base as unknown as ComponentType<any>) : transformAs(as ?? 'div');

    return <Component ref={ref} {...restProps} />;
  }),
  {
    shouldForwardProp,
  }
)(interpolation);
