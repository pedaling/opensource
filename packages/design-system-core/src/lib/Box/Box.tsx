import type { ComponentType, ReactElement } from 'react';
import { forwardRef } from 'react';
import styled from '@emotion/styled';
import { interpolation } from '../interpolation';
import type { BoxProps } from './BoxProps';
import { shouldForwardProp } from './BoxProps';

export const Box = styled(
  forwardRef<HTMLDivElement, BoxProps>(({ as, base, ...restProps }, ref) => {
    const Component = (base as ComponentType) ?? as ?? 'div';

    return <Component ref={ref} {...restProps} />;
  }),
  {
    shouldForwardProp,
  }
)(interpolation) as <
  BaseComponent extends ComponentType,
  ElementName extends keyof JSX.IntrinsicElements,
  ElementProps extends JSX.IntrinsicElements[ElementName]
>(
  props: BoxProps<BaseComponent, ElementName, ElementProps>
) => ReactElement;
