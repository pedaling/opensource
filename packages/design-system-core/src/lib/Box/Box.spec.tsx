import type { ReactRenderer } from '@class101/design-system-utils';
import { createReactRenderer } from '@class101/design-system-utils';
import { ThemeProvider } from '../ThemeProvider';
import { Box } from './Box';

describe('<Box />', () => {
  const { render } = createReactRenderer();
  let renderer: ReactRenderer;

  describe('as prop', () => {
    describe('when empty box created', () => {
      beforeEach(() => {
        renderer = render(<Box />);
      });

      it('div element created', () => {
        expect(renderer.container.querySelector('div')).toBeTruthy();
      });

      it('match snapshot', () => {
        expect(renderer.container).toMatchSnapshot();
      });
    });

    describe('when span box created', () => {
      beforeEach(() => {
        renderer = render(<Box as="span" />);
      });

      it('span element created', () => {
        expect(renderer.container.querySelector('span')).toBeTruthy();
      });

      it('match snapshot', () => {
        expect(renderer.container).toMatchSnapshot();
      });
    });
  });

  describe('base prop', () => {
    describe('when function component box created', () => {
      beforeEach(() => {
        const Component = (props: any) => (
          <span {...props}>
            <div />
          </span>
        );

        renderer = render(<Box base={Component} id="component" />);
      });

      it('custom element created', () => {
        expect(renderer.container.querySelector('span')).toBeTruthy();

        expect(renderer.container.querySelector('span > div')).toBeTruthy();
      });

      it('id prop injected', () => {
        expect(renderer.container.querySelector('span')?.id).toBe('component');
      });

      it('match snapshot', () => {
        expect(renderer.container).toMatchSnapshot();
      });
    });
  });

  describe('default style', () => {
    describe('when styled box created', () => {
      beforeEach(() => {
        renderer = render(<Box width={300} height={300} backgroundColor="red" />);
      });

      it('css apply on div', () => {
        expect(renderer.container.querySelector('div')).toHaveStyleRule('width', '300px');

        expect(renderer.container.querySelector('div')).toHaveStyleRule('height', '300px');

        expect(renderer.container.querySelector('div')).toHaveStyleRule('background-color', 'red');
      });

      it('match snapshot', () => {
        expect(renderer.container).toMatchSnapshot();
      });
    });

    describe('when themed box created', () => {
      beforeEach(() => {
        renderer = render(
          <ThemeProvider
            theme={{ breakpoints: [200, 400], colors: { light: { primary: 'orange', informative: 'black' } } }}
          >
            <Box width={[300, 400]} height={[200, 300]} backgroundColor={['primary', 'informative']} />
          </ThemeProvider>
        );
      });

      it('default style apply on div', () => {
        expect(renderer.container.querySelector('div')).toHaveStyleRule('width', '300px');

        expect(renderer.container.querySelector('div')).toHaveStyleRule('height', '200px');

        expect(renderer.container.querySelector('div')).toHaveStyleRule('background-color', 'orange');
      });

      it('responsive style apply on div', () => {
        expect(renderer.container.querySelector('div')).toHaveStyleRule('width', '400px', {
          media: '@media screen and (min-width: 200px)',
        });

        expect(renderer.container.querySelector('div')).toHaveStyleRule('height', '300px', {
          media: '@media screen and (min-width: 200px)',
        });

        expect(renderer.container.querySelector('div')).toHaveStyleRule('background-color', 'black', {
          media: '@media screen and (min-width: 200px)',
        });
      });

      it('match snapshot', () => {
        expect(renderer.container).toMatchSnapshot();
      });
    });
  });
});
