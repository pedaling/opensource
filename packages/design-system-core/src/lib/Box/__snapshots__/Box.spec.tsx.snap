// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Box /> as prop when empty box created match snapshot 1`] = `
<div>
  <div
    class="emotion-test-renderer-0"
  />
</div>
`;

exports[`<Box /> as prop when span box created match snapshot 1`] = `
<div>
  <span
    class="emotion-test-renderer-0"
  />
</div>
`;

exports[`<Box /> base prop when function component box created match snapshot 1`] = `
<div>
  <span
    class="emotion-test-renderer-0"
    id="component"
  >
    <div />
  </span>
</div>
`;
